src/python/pants/engine/query.py

```diff
@@ -0,0 +1,250 @@
+
+from pants.build_graph.address import Address
+from pants.engine.collection import Collection
+from pants.engine.internals.graph import Owners, OwnersRequest
+from pants.engine.rules import RootRule, rule
+from pants.engine.selectors import Get
+from pants.engine.unions import UnionMembership, UnionRule, union
+from pants.scm.git import Git
+from pants.scm.subsystems.changed import (
+    ChangedFiles,
+    ChangedFilesRequest,
+    ChangedOptions,
+    ChangedAddresses,
+    ChangedRequest,
+    DependeesOption,
+)
+from pants.util.meta import classproperty
+from pants.util.strutil import safe_shlex_split
+
+
+@union
+class QueryComponent(ABC):
+
+    @classproperty
+    @abstractmethod
+    def function_name(cls):
+        """The initial argument of a shlexed query expression.
+
+        If the user provides --query='<name> <args...>' on the command line, and `<name>` matches this
+        property, the .parse_from_args() method is invoked with `<args...>` (shlexed, so split by
+        spaces).
+        """
+
+    @classmethod
+    @abstractmethod
+    def parse_from_args(cls, *args):
+        """Create an instance of this class from variadic positional string arguments.
+
+        This method should raise an error if the args are incorrect or invalid.
+        """
+
+
+class QueryAddresses(Collection[Address]):
+    pass
+
+
+@dataclass(frozen=True)
+class OwnerOf(QueryComponent):
+    files: Tuple[str]
+
+    function_name = 'owner_of'
+
+    @classmethod
+    def parse_from_args(cls, *args):
+        return cls(files=tuple([str(f) for f in args]))
+
+
+@rule
+async def owner_of_request(owner_of: OwnerOf) -> QueryAddresses:
+    request = OwnersRequest(sources=owner_of.files, do_not_generate_subtargets=True)
+    owners = await Get(Owners, OwnersRequest, request)
+    return QueryAddresses(owners)
+
+
+@dataclass(frozen=True)
+class ChangesSince(QueryComponent):
+    since: str
+    dependees: DependeesOption
+
+    function_name = 'since'
+
+    @classmethod
+    def parse_from_args(cls, since, dependees=DependeesOption.NONE):
+        return cls(since=str(since),
+                   dependees=DependeesOption(dependees))
+
+
+@rule
+async def since_request(
+        git: Git,
+        since: ChangesSince,
+) -> QueryAddresses:
+    changed_options = ChangedOptions(
+        since=since.since,
+        diffspec=None,
+        dependees=since.dependees,
+    )
+    changed_files = await Get(ChangedFiles, ChangedFilesRequest(changed_options, git))
+    changed = await Get(ChangedAddresses, ChangedRequest(
+        sources=tuple(changed_files),
+        dependees=changed_options.dependees,
+        do_not_generate_subtargets=True,
+    ))
+    return QueryAddresses(changed)
+
+
+@dataclass(frozen=True)
+class ChangesForDiffspec(QueryComponent):
+    diffspec: str
+    dependees: DependeesOption
+
+    function_name = 'changes_for_diffspec'
+
+    @classmethod
+    def parse_from_args(cls, diffspec, dependees=DependeesOption.NONE):
```

Comment:
I'm a little confused how the user would specify `dependees`. Could you please add an example to the PR description?